src/main/java/org/zalando/nakadi/service/CursorsService.java

```diff
@@ -144,62 +165,86 @@ private boolean isActiveSession(final String subscriptionId, final String stream
         }
     }
 
-    private boolean processCursor(final String subscriptionId, final SubscriptionCursor cursor)
-            throws InternalNakadiException, NoSuchEventTypeException, InvalidCursorException,
-            ServiceUnavailableException, NoSuchSubscriptionException {
+    private List<Boolean> processPartitionCursors(final String subscriptionId, final List<SubscriptionCursor> cursors,
+                                                  final EventTypePartition eventTypePartition)
+            throws InternalNakadiException, NoSuchEventTypeException, ServiceUnavailableException,
+            InvalidCursorException {
+
+        final EventType eventType = eventTypeRepository.findByName(eventTypePartition.getEventType());
+
+        try {
+            final List<NakadiCursor> nakadiCursors = cursors.stream()
+                    .map(cursor -> {
+                        SubscriptionCursor cursorToProcess = cursor;
+                        if (Cursor.BEFORE_OLDEST_OFFSET.equals(cursor.getOffset())) {
+                            cursorToProcess = new SubscriptionCursor(cursor.getPartition(), "-1", cursor.getEventType(),
+                                    cursor.getCursorToken());
+                        }
+                        final NakadiCursor nakadiCursor = new NakadiCursor(
+                                eventType.getTopic(),
+                                cursorToProcess.getPartition(),
+                                cursorToProcess.getOffset());
+                        try {
+                            topicRepository.validateCommitCursor(nakadiCursor);
```

Comment:
Because this method is one that expects a lot of load, I think that it's better not to call kafka several times for list of partitions (here we call listPartitions() each time when validating nakadiCursor). And this call is actually not needed, because partition is correct (as it is present in zk node). So in the end - we are doing many calls that are not needed.